public/app/plugins/datasource/tempo/traceql/QueryEditor.tsx

```diff
@@ -1,25 +1,42 @@
+import { css } from '@emotion/css';
 import { defaults } from 'lodash';
-import React from 'react';
+import React, { useState } from 'react';
 
 import { QueryEditorProps } from '@grafana/data';
+import { useStyles2 } from '@grafana/ui';
 
 import { TempoDatasource, TempoQuery } from '../datasource';
 import { defaultQuery, MyDataSourceOptions } from '../types';
 
+import { TempoQueryBuilderOptions } from './TempoQueryBuilderOptions';
 import { TraceQLEditor } from './TraceQLEditor';
 
 type Props = QueryEditorProps<TempoDatasource, TempoQuery, MyDataSourceOptions>;
 
 export function QueryEditor(props: Props) {
-  function onEditorChange(value: string) {
-    props.onChange({ ...props.query, query: value });
-  }
+  const [query, setQuery] = useState(defaults(props.query, defaultQuery));
+  const styles = useStyles2(getStyles);
 
-  let query = defaults(props.query, defaultQuery);
+  const onLimitChange = (e: React.FormEvent<HTMLInputElement>) => {
+    setQuery({ ...query, limit: parseInt(e.currentTarget.value, 10) });
+  };
+
+  const onEditorChange = (value: string) => {
+    props.onChange({ ...query, query: value });
+  };
 
   return (
-    <div className="gf-form">
+    <>
       <TraceQLEditor value={query.query} onChange={onEditorChange} datasource={props.datasource} />
-    </div>
+      <div className={styles.optionsContainer}>
+        <TempoQueryBuilderOptions query={query} onLimitChange={onLimitChange} />
```

Comment:
It's better to pass a generic `onChange` function to the Options component and let it manage the update of each individual field inside the component. To the `QueryEditor` it shouldn't matter what the implementation of a limit change is, it should only matter that the query is updated, so it should pass the implementation responsibility closer to where the limit field (in this case) is declared. 

If you start laying some groundwork for the future and prepare the code structure to be ready to handle 5/10/15 more fields without it being overwhelming to the reader, you'll save some other poor person a ton of refactoring down the road.

Reply:
I am now sending the `onChange` function to the `TempoQueryBuilderOptions` component and an `onLimitChange` function has been created in the aforementioned component which uses the `onChange` function to update the `limit` property of the query object. See commit https://github.com/grafana/grafana/pull/54435/commits/79c99554d7b079a1e221e4d03fd0a929553ed8aa.